### PR TITLE
Mask the shift count during constant folding to imitate the X86 behavior when unspecified.

### LINKED ISSUE #38720: Make constant folding of `<<` and `>>` host platform independent
Constant folding of `<<` and `>>` in the C# compiler depends on the (undefined) behavior of the underlying platform when the shift amount is greater than the size of the shifted value.  We should mask the shift amount to get deterministic behavior matching the X86 behavior.

Here is the offending source:
``` c#
                case BinaryOperatorKind.IntLeftShift:
                    return valueLeft.Int32Value << valueRight.Int32Value;
                case BinaryOperatorKind.LongLeftShift:
                    return valueLeft.Int64Value << valueRight.Int32Value;
                case BinaryOperatorKind.UIntLeftShift:
                    return valueLeft.UInt32Value << valueRight.Int32Value;
                case BinaryOperatorKind.ULongLeftShift:
                    return valueLeft.UInt64Value << valueRight.Int32Value;
                case BinaryOperatorKind.IntRightShift:
                    return valueLeft.Int32Value >> valueRight.Int32Value;
                case BinaryOperatorKind.LongRightShift:
                    return valueLeft.Int64Value >> valueRight.Int32Value;
                case BinaryOperatorKind.UIntRightShift:
                    return valueLeft.UInt32Value >> valueRight.Int32Value;
                case BinaryOperatorKind.ULongRightShift:
                    return valueLeft.UInt64Value >> valueRight.Int32Value;
```

### PR DESCRIPTION
**Pull request by gafter**
Fixes #38720

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs [hunk 1]
```diff
@@ -2443,6 +2443,324 @@ void M()
             Assert.Equal(expected, actual);
         }
 
+        [Fact]
+        public void TestUnspecifiedShifts()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs]: >TestUnspecifiedShifts [](start = 20, length = 21)

Does VB have the same problem? Please add a test.

Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs]: VB already has masking for its shift operators and tests.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 2]
```diff
@@ -1717,21 +1717,21 @@ private static object FoldNeverOverflowBinaryOperators(BinaryOperatorKind kind,
                 case BinaryOperatorKind.FloatRemainder:
                     return valueLeft.SingleValue % valueRight.SingleValue;
                 case BinaryOperatorKind.IntLeftShift:
-                    return valueLeft.Int32Value << valueRight.Int32Value;
+                    return valueLeft.Int32Value << (valueRight.Int32Value % 32);
```
Comment (mikedn) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: I'm confused. Why is this needed given that the C# spec says in https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#shift-operators

> When the type of x is int or uint, the shift count is given by the low-order five bits of count. In other words, the shift count is computed from count & 0x1F.

Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: Excellent question.  It shouldn't be needed.  I was looking at the ECMA specification for the CLR, which makes it undefined behavior!  Since the C# compiler does not produce the required mask operation before shifting, we are producing incorrect code!  I'll investigate further.  We may need to change the CLR specification to have similar language for the shift instructions.

**Comment by gafter** [comment]
As @mikedn pointed out in https://github.com/dotnet/roslyn/pull/38729#discussion_r325794031, the issue described in https://github.com/dotnet/roslyn/issues/38720 does not exist.  The C# compiler produces code to mask the shift amount to ensure correct behavior on the CLR.  See https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACBmAAhwEYA2IgJgIGECBvAWACgDWjCTy0A7GAgWQAUPPmiQERBAFYBKAgF4AfBIIAeVdIDczAL5A= for a demonstration of this fact.

**Comment by gafter** [comment]
Reopening to keep the test.

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 2 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs]: ok, we do emit a mask operation.
